### PR TITLE
RichText: convert HTML formatting whitespace to spaces

### LINKED ISSUE #11588: Operation on text randomly merges words
**Describe the bug**
When interacting (bolding, inserting links) with text block that has pasted content with prior formatting, block will randomly merge a few words in the text.

**To Reproduce**
1. Copy & paste text into the text block directly from word editor (LibreOffice in my case)
2. Bold a word
3. Enjoy the bug

**Expected behavior**
A few worlds in the text block will be merged.

**Screenshots**
BEFORE:
zrzut ekranu 2018-11-07 o 15 16 23
AFTER:
zrzut ekranu 2018-11-07 o 15 16 31

**Desktop (please complete the following information):**
 - OS: macOS
 - Browser: Chrome
 - Version: 70.0.3538.77

**Additional context**
- Present in latest Gutenberg release
- Avoidable by pasting and matching the style

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Fixes #11588, #5872 and #7474.

When creating a rich text value from text nodes, this branch reduces any sequence of new lines and tabs to one space. This is needed because any sequence of new lines and tabs is used to format HTML, not as content. The browser will only display it as one space, if it is used in between words.

Previously we were just removing any line breaks, which would normally still be displayed by the browser as a space.

It also visualises and leading and trailing spaces, and and consecutive spaces elsewhere. This prevents buggy behaviour in different browsers when there is only one spaces, and prevent e.g. Chrome from inserting non breaking spaces and Firefox a trailing line break to visualise a trailing space when typing.

## How has this been tested?

Edit a paragraph as HTML. Remove a space and insert a line break (use ENTER). Edit visually. You should see a space where you inserted the line break. Edit the content. Edit as HTML. The line break should be replaced by a space.

Previously the line break would have been removed.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by designsimply** [comment]
I re-tested the steps at https://github.com/WordPress/gutenberg/issues/11588#issuecomment-439094181 using the `try/rich-text-newlines-to-space` branch and found that no spaces went missing after copying content from a locally installed word processing app into Gutenberg.

I also tested the following steps:

1. Edit a paragraph as HTML. 
1. Remove a space and insert a line break (use ENTER). 
1. Edit visually. 
1. You should see a space where you inserted the line break. 
1. Edit as HTML. 
1. The line break should be replaced by a space.

Result: after using "Edit as HTML" and replacing a space with a line break and then switching back to "Edit visually," I could see a space where I had inserted a line break. After switching back to "Edit as HTML" the line break was still there and was not replaced by a space (as described in step 7).

screen shot 2018-11-25 at 4 16 35 pm
<sup>Seen at http://localhost:8888/wp-admin/post.php?post=24&action=edit using WordPress 4.9.8 and Gutenberg `0431abae0` (`try/rich-text-newlines-to-space` branch) with Firefox 63.0.3 on macOS 10.13.6.</sup>

**Comment by iseulde** [comment]
@designsimply If you switch between visual and HTML mode without editing the content, no content will be saved, so it's normal that the line break stays there if the content in unchanged.

**Comment by Haldaug** [comment]
This resolved the bug https://github.com/WordPress/gutenberg/issues/11588 for me. Nice work, @iseulde.

**Comment by nosolosw** [comment]
This is what I've tested (both in chrome 70 and firefox 63, Linux OS). This seems ready.

## https://github.com/WordPress/gutenberg/issues/7474 is fixed

Test 1:

* Type "Hey Jude, ". A `nbsp` characters no longer gets added in chrome. Chrome still shows a space, but it is non-breaking.

Test 2:

* Type "Hey Jude, "
* Paste "Don't make it bad."
* A `nbsp` character is no longer inserted in the saved content.

Test 3:

* Type a space " ".
* Paste an embeddable link, such as https://www.youtube.com/watch?v=A_MjCqQoLLA
* In the editor, the embed isn't expanded, but after saving the post, the front-end does show that the embed was created and the iframe correctly added. Previously, it rendered a `nbsp` and the link.

## https://github.com/WordPress/gutenberg/issues/5872

I could only repro this one in part. This is what I've tried:

* Add a paragraph or heading with "hey Jude"
* Go to Code editor mode and add a new line (using `ENTER) and then "Don't let me down"
* Go to visual mode: new line is converted to space.
* I save and publish the post and the front-end also shows the space.

I've also noticed that after saving the post, if I edit that paragraph in the code editor mode, I still see the line break.

## https://github.com/WordPress/gutenberg/issues/11588

What I've tested:

* Copy Hey Jude lyric from Libreoffice.
* The first space between two words is no longer deleted.

**Comment by nosolosw** [comment]
I'm not very familiar with this part of the code, but I'd like us to spent some time addaing some e2e or unit tests with the steps I've reported [above](https://github.com/WordPress/gutenberg/pull/12166#issuecomment-441715210). It does seem like easily reproducible with a test and also easily missed in the future if we don't have them.

I can take a look at adding them tomorrow.

**Comment by iseulde** [comment]
Last time I checked pasting didn't work in puppeteer, but maybe this changed.

**Comment by tofumatt** [comment]
Copy/pasting in Puppeteer still, from what I can tell, doesn't work. I think it's because it's more of an OS-level interaction and not a browser one.

Last time I tried to write an E2E test with copy/paste I went down an hour-long rabbit-hole and then ended up stopping 🤷‍♂️

File: packages/editor/src/components/rich-text/style.scss [hunk 1]
```diff
@@ -8,6 +8,16 @@
...
+	// In HTML, leading and trailing spaces are not visible, and multiple spaces
+	// elsewhere are visually reduced to one space. This rule prevents spaces
+	// from collapsing so all space is visible in the editor and can be removed.
+	// It also prevents some browsers from inserting non-breaking spaces at the
+	// end of a line to prevent the space from visually disappearing. Sometimes
+	// these non breaking spaces can linger in the editor causing unwanted non
+	// breaking spaces in between words. If also prevent Firefox from inserting
+	// a trailing `br` node to visualise any trailing space, causing the element
+	// to be saved.
+	white-space: pre-wrap;
```
Comment (nosolosw) [inline packages/editor/src/components/rich-text/style.scss]: `pre-wrap` will preserve newlines, spaces, and tabs. But in `filterStringComplete` we're only replacing newlines and tabs with spaces. As a consequence of this change, the editor will now show those consecutive spaces.

Before this change:

peek 2018-11-27 19-21

After this change:

peek 2018-11-27 19-29

Is that a concern?

File: packages/rich-text/src/create.js [hunk 2]
```diff
@@ -297,11 +297,10 @@ function createFromElement( {
...
 	const length = element.childNodes.length;
 
-	// Remove any line breaks in text nodes. They are not content, but used to
-	// format the HTML. Line breaks in HTML are stored as BR elements.
-	// See https://www.w3.org/TR/html5/syntax.html#newlines.
 	const filterStringComplete = ( string ) => {
-		string = string.replace( /[\r\n]/g, '' );
+		// Reduce any whitespace used for HTML formatting to one space
+		// character, because it will also be displayed as such by the browser.
+		string = string.replace( /[\n\r\t]+/g, ' ' );
```
Comment (nosolosw) [inline packages/rich-text/src/create.js]: Would it make sense to replace `\r` and `\n` with no spaces as it was before, and any number of tabs with one space?

File: packages/editor/src/components/rich-text/style.scss (hunk 1 above)
Comment (nosolosw) [reply packages/editor/src/components/rich-text/style.scss]: For completeness, here's the same action in visual mode. No visual change, although under the hood the HTML doesn't add `nbsp`.

Before this change:

peek 2018-11-27 19-23

After this change:

peek 2018-11-27 19-28

I've noticed that with this change applied the leading spaces in the first line collapse before the post is reloaded.

Comment (iseulde) [reply packages/editor/src/components/rich-text/style.scss]: I think that’s fine?

File: packages/rich-text/src/create.js (hunk 2 above)
Comment (iseulde) [reply packages/rich-text/src/create.js]: No, one of the issues this or fixes is exactly replacing line breaks with a space.

File: packages/editor/src/components/rich-text/style.scss (hunk 1 above)
Comment (iseulde) [reply packages/editor/src/components/rich-text/style.scss]: In the rare case that the user want to have those spaces visible on the front end, I think they should insert those non breaking spaces themselves.

**Review by oandregal — APPROVED** [review]
Now that I've familiarized myself more with the rich-text library, this looks fine code-wise. I couldn't find a way to break this with the changes proposed :)

**Comment by iseulde** [comment]
@mcsf It shouldn't affect the fix. In Twenty Nineteen we're replacing a font tile with another, because Hoefler Text has a weird design for a non breaking space. It is true that Chrome will now no longer insert trailing non breaking spaces, so if the fix were to be removed, typing would be fine. But other non breaking spaces would still be displayed too wide.

**Comment by mcsf** [comment]
Thanks for having a look.

**Comment by peteschiebel** [comment]
This looks to have been merged into GB 4.7, which was originally intended to be part of WordPress 5.0.1, but then maybe shifted to GB 4.8/ WP 5.0.2 because of the quick 5.0.1 release...
1) Has this made it into the core Block Editor?
2) Is there a handy way to check if a specific GB PR has made it into the latest version of WordPress?
Thanks for the hard work!

**Pull request merged**